bitvm/src/bn254/g2.rs

```diff
@@ -0,0 +1,2186 @@
+
+    }
+}
+
+
+#[derive(Debug, Default, Clone)]
+pub struct ScriptContext<F: ark_ff::Field> {
+    pub inputs: Vec<F>,
+    pub outputs: Vec<F>,
+    pub auxiliary: Vec<usize>,
+}
+
+#[derive(Debug, Clone, Default)]
+pub struct SplitScript {
+    pub script: ScriptBuf,
+    pub input_len: u32,
+    pub output_len: u32,
+}
+
+pub struct PairingNative;
+
+impl PairingNative {
+    // Reference: https://github.com/arkworks-rs/algebra/blob/master/curves/bn254/src/curves/g2.rs#L59
+    // https://github.com/BitVM/BitVM/issues/109
+    pub fn witness_g2_subgroup_check(
+        point: &ark_bn254::g2::G2Affine,
+        constants: [ark_bn254::Fq2; 2],
+        scalar_bit: Vec<bool>,
+    ) -> (bool, Vec<ScriptContext<ark_bn254::Fq2>>) {
+        let mut script_contexts = vec![];
+
+        let mut script_context = ScriptContext::default();
+
+        // Maps (x,y) -> (x^p * (u+9)^((p-1)/3), y^p * (u+9)^((p-1)/2))
+        script_context
+            .inputs
+            .push(point.clone().x().unwrap().to_owned());
+        script_context
+            .inputs
+            .push(point.clone().y().unwrap().to_owned());
+
+        let mut p_times_point = *point;
+        p_times_point.x.frobenius_map_in_place(1);
+        p_times_point.y.frobenius_map_in_place(1);
+
+        p_times_point.x *= constants[0];
+        p_times_point.y *= constants[1];
+
+        script_context
+            .outputs
+            .push(p_times_point.clone().x().unwrap().to_owned());
+        script_context
+            .outputs
+            .push(p_times_point.clone().y().unwrap().to_owned());
+
+        script_contexts.push(script_context);
+
+        let (x_times_point, witness) = Self::witness_split_scalar_mul_g2(point, &scalar_bit);
+
+        assert_eq!(p_times_point, x_times_point);
+
+        script_contexts.extend(witness);
+
+        (true, script_contexts)
+    }
+    
+    pub fn witness_split_scalar_mul_g2(
+        base: &ark_bn254::G2Affine,
+        scalar: &[bool],
+    ) -> (ark_bn254::G2Affine, Vec<ScriptContext<ark_bn254::Fq2>>) {
+        let res = base.to_owned();
+        let mut tmp = base.to_owned();
+
+        let mut script_contexts = vec![];
+
+        for b in scalar.iter().skip(1) {
+
+            let (lambda, miu, res_x, res_y) = PairingNative::line_double_g2(&tmp);
+
+            let mut script_context = ScriptContext::default();
+
+            script_context.inputs.push(lambda);
+            script_context.inputs.push(miu);
+            script_context.inputs.push(tmp.x().unwrap().to_owned());
+            script_context.inputs.push(tmp.y().unwrap().to_owned());
+            script_context.outputs.push(res_x);
+            script_context.outputs.push(res_y);
+
+            tmp = tmp.add(tmp).into_affine();
+
+            assert_eq!(res_x, tmp.x().unwrap().clone());
+            assert_eq!(res_y, tmp.y().unwrap().clone());
+
+            // ecc_double_add_data_set.push(ecc_double_add_data_tmp);
```

Comment:
Comment seems outdated. There is no `ecc_double_add_data_set` as far as I can see.